contrib/99designs/gqlgen/tracer.go

```diff
@@ -0,0 +1,148 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-2019 Datadog, Inc.
+
+// Package gqlgen contains an implementation of a gqlgen tracer, and functions to construct and configure the tracer.
+// The tracer can be passed to the gqlgen handler (see package github.com/99designs/gqlgen/handler)
+package gqlgen
+
+import (
+	"context"
+	"fmt"
+	"math"
+
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/ext"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
+
+	"github.com/99designs/gqlgen/graphql"
+)
+
+const (
+	tagResolverObject = "resolver.object"
+	tagResolverField  = "resolver.field"
+)
+
+type gqlTracer struct {
+	cfg *config
+}
+
+// New creates an a graphql.Tracer instance that can be passed to a gqlgen handler.
+// Options can be passed in for further configuration.
+func New(opts ...Option) graphql.Tracer {
+	t := &gqlTracer{cfg: &config{}}
+	defaults(t.cfg)
+	for _, o := range opts {
+		o(t.cfg)
+	}
+	return t
+}
+
+// gqlTracer implements the graphql.Tracer interface.
+func (t *gqlTracer) StartOperationParsing(ctx context.Context) context.Context {
+	return ctx
+}
+
+// gqlTracer implements the graphql.Tracer interface.
+func (t *gqlTracer) EndOperationParsing(ctx context.Context) {
+	// not implemented
+}
+
+// gqlTracer implements the graphql.Tracer interface.
+func (t *gqlTracer) StartOperationValidation(ctx context.Context) context.Context {
+	return ctx
+}
+
+// gqlTracer implements the graphql.Tracer interface.
+func (t *gqlTracer) EndOperationValidation(ctx context.Context) {
+	// not implemented
+}
+
+func (t *gqlTracer) StartOperationExecution(ctx context.Context) context.Context {
+	opts := []ddtrace.StartSpanOption{
+		tracer.SpanType(ext.SpanTypeGraphql),
+		tracer.ServiceName(t.cfg.serviceName),
+	}
+	if !math.IsNaN(t.cfg.analyticsRate) {
+		opts = append(opts, tracer.Tag(ext.EventSampleRate, t.cfg.analyticsRate))
+	}
+	rctx := graphql.GetRequestContext(ctx)
+	if rctx != nil {
+		opts = append(opts, tracer.Tag("query", rctx.RawQuery))
+		if rctx.OperationName != "" {
+			opts = append(opts, tracer.ResourceName(rctx.OperationName))
+		}
+		if rctx.ComplexityLimit > 0 {
+			opts = append(opts,
+				tracer.Tag("complexityLimit", rctx.ComplexityLimit),
+				tracer.Tag("operationComplexity", rctx.OperationComplexity),
+			)
+		}
+		for key, val := range rctx.Variables {
+			opts = append(opts,
+				tracer.Tag(fmt.Sprintf("variables.%s", key), fmt.Sprintf("%+v", val)),
+			)
+		}
+	}
+
+	if s, ok := tracer.SpanFromContext(ctx); ok {
+		opts = append(opts, tracer.ChildOf(s.Context()))
+	}
+	_, ctx = tracer.StartSpanFromContext(ctx, "gqlgen.operation", opts...)
+	return ctx
+}
+
+func (t *gqlTracer) StartFieldExecution(ctx context.Context, field graphql.CollectedField) context.Context {
```

Comment:
We are not making use of the `field` argument at all here. It contains a ton of information. I think the [Opencensus implementation](https://github.com/99designs/gqlgen-contrib/blob/master/gqlopencensus/tracer.go#L67) makes a lot of sense. We should add the same information as they do, as soon as we have it.